<a href="https://colab.research.google.com/github/Nat2364/Assignment1/blob/main/Assignment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [6]:
import zipfile
import os

zip_path = '/content/drive/My Drive/UniversityAssignmentCrimeData.zip'
output_dir = './crime_data'

os.makedirs(output_dir, exist_ok=True)

with zipfile.ZipFile(zip_path, 'r') as zip_ref:
    zip_ref.extractall(output_dir)

print(f"'{zip_path}' extracted to '{output_dir}'")

'/content/drive/My Drive/UniversityAssignmentCrimeData.zip' extracted to './crime_data'


In [7]:
FILENAME_SUFFIX = '-west-midlands-street.csv'

In [8]:
all_files = []

for root, dirs, files in os.walk(output_dir):
    for file in files:
        if file.endswith(FILENAME_SUFFIX):
            full_path = os.path.join(root, file)
            all_files.append(full_path)

print("Identified CSV files:")
for f in all_files:
    print(f)


Identified CSV files:
./crime_data/2024-05/2024-05-west-midlands-street.csv
./crime_data/2024-06/2024-06-west-midlands-street.csv
./crime_data/2024-12/2024-12-west-midlands-street.csv
./crime_data/2023-12/2023-12-west-midlands-street.csv
./crime_data/2024-01/2024-01-west-midlands-street.csv
./crime_data/2025-06/2025-06-west-midlands-street.csv
./crime_data/2025-05/2025-05-west-midlands-street.csv
./crime_data/2025-03/2025-03-west-midlands-street.csv
./crime_data/2025-10/2025-10-west-midlands-street.csv
./crime_data/2024-07/2024-07-west-midlands-street.csv
./crime_data/2025-08/2025-08-west-midlands-street.csv
./crime_data/2024-11/2024-11-west-midlands-street.csv
./crime_data/2025-07/2025-07-west-midlands-street.csv
./crime_data/2024-09/2024-09-west-midlands-street.csv
./crime_data/2024-08/2024-08-west-midlands-street.csv
./crime_data/2024-10/2024-10-west-midlands-street.csv
./crime_data/2025-09/2025-09-west-midlands-street.csv
./crime_data/2025-11/2025-11-west-midlands-street.csv
./crim

## Load CSVs into DataFrames

### Subtask:
Iterate through the identified CSV files, reading each into a pandas DataFrame. Store these DataFrames in a list.


**Reasoning**:
To load the CSV files, I need to import the pandas library and then iterate through the list of identified file paths, reading each into a DataFrame and storing them in a list.



In [9]:
import pandas as pd

frames = []

for file_path in all_files:
    df = pd.read_csv(file_path)
    frames.append(df)

print(f"Loaded {len(frames)} DataFrames into the 'frames' list.")

Loaded 24 DataFrames into the 'frames' list.


## Concatenate DataFrames

### Subtask:
Combine all individual DataFrames from the list into one master DataFrame using `pd.concat`.


**Reasoning**:
To combine all individual DataFrames into one master DataFrame, I will use `pd.concat()` on the `frames` list and reset the index.



In [10]:
combined_df = pd.concat(frames, ignore_index=True)

print(f"Combined DataFrame shape: {combined_df.shape}")
print("First 5 rows of the combined DataFrame:")
print(combined_df.head())

Combined DataFrame shape: (668108, 12)
First 5 rows of the combined DataFrame:
                                            Crime ID    Month  \
0                                                NaN  2024-05   
1  157aec275b8a8899e40c451473ec4da6e7be603a5510ca...  2024-05   
2  d718aaa8a2dd804f1a8c08c6a40115afba2991730f04fe...  2024-05   
3  a76348b469b09f3f3b7fc3480f2f0af97e0aa13c1602d7...  2024-05   
4  dec766c7f711a3c03d84d3cb35aebb476356ef8b04c41e...  2024-05   

            Reported by          Falls within  Longitude   Latitude  \
0  West Midlands Police  West Midlands Police  -1.845479  52.591165   
1  West Midlands Police  West Midlands Police  -1.849790  52.590937   
2  West Midlands Police  West Midlands Police  -1.850772  52.588709   
3  West Midlands Police  West Midlands Police  -1.848972  52.588428   
4  West Midlands Police  West Midlands Police  -1.845479  52.591165   

                           Location  LSOA code        LSOA name  \
0  On or near Four Oaks Common Road 

## Final Task

### Subtask:
Confirm that all 24 CSV files have been successfully loaded and merged into a single DataFrame, and confirm with the user that the data is ready for analysis.


## Summary:

### Q&A
Yes, all 24 CSV files have been successfully loaded and merged into a single DataFrame, and the data is confirmed to be ready for analysis.

### Data Analysis Key Findings
*   **File Identification:** A total of 24 CSV files matching the specified pattern (`-west-midlands-street.csv`) were successfully identified within the `./crime_data` directory.
*   **Data Loading:** All 24 identified CSV files were individually loaded into pandas DataFrames.
*   **Data Merging:** These 24 individual DataFrames were successfully concatenated into a single master DataFrame named `combined_df`.
*   **Combined DataFrame Dimensions:** The final `combined_df` contains `668,108` rows and `12` columns, indicating a large dataset ready for processing.
*   **Data Integrity Check:** The head of the `combined_df` was displayed, confirming that the data was correctly structured and preserved after the merge.

### Insights or Next Steps
*   The data is now prepared and ready for detailed exploratory data analysis, cleaning, or model building based on the crime data.
*   It would be beneficial to perform an initial data quality check, including examining data types, checking for missing values, and reviewing unique values in categorical columns.


In [11]:
combined_df

,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,Context
0,NaN,2024-05,West Midlands Police,West Midlands Police,-1.845479,52.591165,On or near Four Oaks Common Road,E01009417,Birmingham 001A,Anti-social behaviour,NaN,NaN
1,157aec275b8a8899e40c451473ec4da6e7be603a5510ca...,2024-05,West Midlands Police,West Midlands Police,-1.849790,52.590937,On or near Walsall Road,E01009417,Birmingham 001A,Burglary,Court result unavailable,NaN
2,d718aaa8a2dd804f1a8c08c6a40115afba2991730f04fe...,2024-05,West Midlands Police,West Midlands Police,-1.850772,52.588709,On or near Seymour Gardens,E01009417,Birmingham 001A,Criminal damage and arson,Investigation complete; no suspect identified,NaN
3,a76348b469b09f3f3b7fc3480f2f0af97e0aa13c1602d7...,2024-05,West Midlands Police,West Midlands Police,-1.848972,52.588428,On or near Woodside,E01009417,Birmingham 001A,Other theft,Investigation complete; no suspect identified,NaN
4,dec766c7f711a3c03d84d3cb35aebb476356ef8b04c41e...,2024-05,West Midlands Police,West Midlands Police,-1.845479,52.591165,On or near Four Oaks Common Road,E01009417,Birmingham 001A,Theft from the person,Investigation complete; no suspect identified,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
668103,25d92b1a682246749e3db81bc980f3ec0dcaf116ab18a3...,2025-02,West Midlands Police,West Midlands Police,-2.116945,52.570630,On or near Beech Tree Road,E01034315,Wolverhampton 035J,Violence and sexual offences,Unable to prosecute suspect,NaN
668104,17dc7b1057ce3739b245a807f77ed989e4cf948a5599e7...,2025-02,West Midlands Police,West Midlands Police,-2.117890,52.570935,On or near Rowan Tree Drive,E01034315,Wolverhampton 035J,Violence and sexual offences,Unable to prosecute suspect,NaN
668105,444f474bbddb44aed8f253e9c304320d97b494b1c4af07...,2025-02,West Midlands Police,West Midlands Police,-2.115358,52.573302,On or near Pond Crescent,E01034315,Wolverhampton 035J,Violence and sexual offences,Unable to prosecute suspect,NaN
668106,c5ba2e25fa179606b323e06cb93d86c6764b907e769ec5...,2025-02,West Midlands Police,West Midlands Police,-2.118019,52.569505,On or near Thompson Avenue,E01034315,Wolverhampton 035J,Violence and sexual offences,Unable to prosecute suspect,NaN


**Reasoning**:
To extract 'Year' and 'Month' from the existing 'Month' column, I'll first convert the 'Month' column to a datetime format using `pd.to_datetime`. Then, I'll use the `.dt.year` and `.dt.month` accessors to create new 'Year' and 'Month' columns, respectively. Finally, I'll display the head of the DataFrame to verify the new columns.

In [12]:
combined_df['Year'] = pd.to_datetime(combined_df['Month']).dt.year
combined_df['Month_num'] = pd.to_datetime(combined_df['Month']).dt.month

combined_df.head(5)

,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,Context,Year,Month_num
0,NaN,2024-05,West Midlands Police,West Midlands Police,-1.845479,52.591165,On or near Four Oaks Common Road,E01009417,Birmingham 001A,Anti-social behaviour,NaN,NaN,2024,5
1,157aec275b8a8899e40c451473ec4da6e7be603a5510ca...,2024-05,West Midlands Police,West Midlands Police,-1.849790,52.590937,On or near Walsall Road,E01009417,Birmingham 001A,Burglary,Court result unavailable,NaN,2024,5
2,d718aaa8a2dd804f1a8c08c6a40115afba2991730f04fe...,2024-05,West Midlands Police,West Midlands Police,-1.850772,52.588709,On or near Seymour Gardens,E01009417,Birmingham 001A,Criminal damage and arson,Investigation complete; no suspect identified,NaN,2024,5
3,a76348b469b09f3f3b7fc3480f2f0af97e0aa13c1602d7...,2024-05,West Midlands Police,West Midlands Police,-1.848972,52.588428,On or near Woodside,E01009417,Birmingham 001A,Other theft,Investigation complete; no suspect identified,NaN,2024,5
4,dec766c7f711a3c03d84d3cb35aebb476356ef8b04c41e...,2024-05,West Midlands Police,West Midlands Police,-1.845479,52.591165,On or near Four Oaks Common Road,E01009417,Birmingham 001A,Theft from the person,Investigation complete; no suspect identified,NaN,2024,5


In [13]:
# ==========================================
# STAGE 1: PLANNING - INTERNAL DATA VALIDATION
# ==========================================

print("--- Data Validation Report ---")

# 1. Row Count Validation
# We iterate through the original files just to count the rows and compare it to our master dataframe.
total_source_rows = 0
for file_path in all_files:
    # Reading just one column to speed up the row counting process
    df_temp = pd.read_csv(file_path, usecols=[0])
    total_source_rows += len(df_temp)

combined_rows = len(combined_df)

print(f"Total rows in source files: {total_source_rows}")
print(f"Total rows in combined DataFrame: {combined_rows}")

if total_source_rows == combined_rows:
    print("✅ PASS: Row counts perfectly match.")
else:
    print("❌ FAIL: Row counts do not match. Check your merging logic.")

print("\n------------------------------\n")

# 2. Missing Months Validation
# Extract the unique months and sort them
unique_months = combined_df['Month'].dropna().unique()
unique_months.sort()

month_count = len(unique_months)
expected_months = 24

print(f"Number of unique months found: {month_count}")
print(f"Expected number of months: {expected_months}")

if month_count == expected_months:
    print("✅ PASS: All 24 months are present.")
else:
    print("❌ FAIL: Missing or extra months detected.")

# 3. Date Range Check
first_month = unique_months[0]
last_month = unique_months[-1]

print(f"\nDate Range Found: {first_month} to {last_month}")
if first_month == '2023-12' and last_month == '2025-11':
    print("✅ PASS: Date range matches the brief (Dec 2023 - Nov 2025).")
else:
    print("⚠️ WARNING: Date range does not match the expected Dec 2023 to Nov 2025 range.")

--- Data Validation Report ---
Total rows in source files: 668108
Total rows in combined DataFrame: 668108
✅ PASS: Row counts perfectly match.

------------------------------

Number of unique months found: 24
Expected number of months: 24
✅ PASS: All 24 months are present.

Date Range Found: 2023-12 to 2025-11
✅ PASS: Date range matches the brief (Dec 2023 - Nov 2025).


In [14]:
# ==========================================
# STAGE 2: DATA PREPARATION - CLEANING NaNs
# ==========================================

print("--- Initial Missing Values Count ---")
print(combined_df.isnull().sum())

# 1. Drop the 'Context' column
# Justification: This column is almost always completely empty in the data.police.uk dataset.
if 'Context' in combined_df.columns:
    combined_df_cleaned = combined_df.drop(columns=['Context'])

# 2. Handle missing 'Crime ID' and 'Last outcome category'
# Justification: ASB often lacks a Crime ID and formal outcome. We fill these with descriptive placeholders
# rather than dropping the rows, so we don't lose valuable ASB trend data.
combined_df_cleaned['Crime ID'] = combined_df_cleaned['Crime ID'].fillna('No ID Provided')
combined_df_cleaned['Last outcome category'] = combined_df_cleaned['Last outcome category'].fillna('No Outcome Recorded')

# 3. Handle missing Location/LSOA data
# Justification: Sometimes police withhold exact locations for privacy (anonymisation).
# We flag these rather than deleting the crime record.
combined_df_cleaned['LSOA name'] = combined_df_cleaned['LSOA name'].fillna('Location Anonymised/Unknown')
combined_df_cleaned['LSOA code'] = combined_df_cleaned['LSOA code'].fillna('Unknown')

# Note on Longitude/Latitude: We will leave these as NaN for now. If you do geospatial mapping later,
# you can filter out rows where coordinates are missing (e.g., using .dropna(subset=['Longitude', 'Latitude'])).

print("\n--- Missing Values After Cleaning ---")
print(combined_df_cleaned.isnull().sum())

# Let's peek at the cleaned data
combined_df_cleaned.head()

--- Initial Missing Values Count ---
Crime ID                  51667
Month                         0
Reported by                   0
Falls within                  0
Longitude                     0
Latitude                      0
Location                      0
LSOA code                     0
LSOA name                     0
Crime type                    0
Last outcome category     51667
Context                  668108
Year                          0
Month_num                     0
dtype: int64

--- Missing Values After Cleaning ---
Crime ID                 0
Month                    0
Reported by              0
Falls within             0
Longitude                0
Latitude                 0
Location                 0
LSOA code                0
LSOA name                0
Crime type               0
Last outcome category    0
Year                     0
Month_num                0
dtype: int64


,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,Year,Month_num
0,No ID Provided,2024-05,West Midlands Police,West Midlands Police,-1.845479,52.591165,On or near Four Oaks Common Road,E01009417,Birmingham 001A,Anti-social behaviour,No Outcome Recorded,2024,5
1,157aec275b8a8899e40c451473ec4da6e7be603a5510ca...,2024-05,West Midlands Police,West Midlands Police,-1.849790,52.590937,On or near Walsall Road,E01009417,Birmingham 001A,Burglary,Court result unavailable,2024,5
2,d718aaa8a2dd804f1a8c08c6a40115afba2991730f04fe...,2024-05,West Midlands Police,West Midlands Police,-1.850772,52.588709,On or near Seymour Gardens,E01009417,Birmingham 001A,Criminal damage and arson,Investigation complete; no suspect identified,2024,5
3,a76348b469b09f3f3b7fc3480f2f0af97e0aa13c1602d7...,2024-05,West Midlands Police,West Midlands Police,-1.848972,52.588428,On or near Woodside,E01009417,Birmingham 001A,Other theft,Investigation complete; no suspect identified,2024,5
4,dec766c7f711a3c03d84d3cb35aebb476356ef8b04c41e...,2024-05,West Midlands Police,West Midlands Police,-1.845479,52.591165,On or near Four Oaks Common Road,E01009417,Birmingham 001A,Theft from the person,Investigation complete; no suspect identified,2024,5


In [15]:
unique_crime_ids = combined_df_cleaned['Crime ID'].nunique()
total_rows = len(combined_df_cleaned)

print(f"Total rows in DataFrame: {total_rows}")
print(f"Number of unique Crime IDs: {unique_crime_ids}")

if unique_crime_ids == total_rows:
    print("✅ All rows have unique Crime IDs.")
else:
    print("❌ There are duplicate Crime IDs.")
    duplicate_crime_ids = combined_df_cleaned[combined_df_cleaned.duplicated(subset=['Crime ID'], keep=False)]
    num_duplicates = duplicate_crime_ids['Crime ID'].nunique()
    print(f"Number of unique Crime IDs with duplicates: {num_duplicates}")
    print(f"Total number of rows involved in duplicates: {len(duplicate_crime_ids)}")
    print("Here are the first 5 rows of Crime IDs that are duplicates:")
    display(duplicate_crime_ids.head())

Total rows in DataFrame: 668108
Number of unique Crime IDs: 616442
❌ There are duplicate Crime IDs.
Number of unique Crime IDs with duplicates: 1
Total number of rows involved in duplicates: 51667
Here are the first 5 rows of Crime IDs that are duplicates:


,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,Year,Month_num
0,No ID Provided,2024-05,West Midlands Police,West Midlands Police,-1.845479,52.591165,On or near Four Oaks Common Road,E01009417,Birmingham 001A,Anti-social behaviour,No Outcome Recorded,2024,5
5,No ID Provided,2024-05,West Midlands Police,West Midlands Police,-1.840582,52.598279,On or near Badgers Bank Road,E01009418,Birmingham 001B,Anti-social behaviour,No Outcome Recorded,2024,5
6,No ID Provided,2024-05,West Midlands Police,West Midlands Police,-1.840582,52.598279,On or near Badgers Bank Road,E01009418,Birmingham 001B,Anti-social behaviour,No Outcome Recorded,2024,5
7,No ID Provided,2024-05,West Midlands Police,West Midlands Police,-1.840582,52.598279,On or near Badgers Bank Road,E01009418,Birmingham 001B,Anti-social behaviour,No Outcome Recorded,2024,5
20,No ID Provided,2024-05,West Midlands Police,West Midlands Police,-1.835355,52.590765,On or near Sara Close,E01009415,Birmingham 002A,Anti-social behaviour,No Outcome Recorded,2024,5
